Replace ASCII control chars with Unicode Control Pictures

estebank opened the pull request:
Replace ASCII control chars like `CR` with Unicode Control Pictures like `␍`:

```
error: bare CR not allowed in doc-comment
  --> $DIR/lex-bare-cr-string-literal-doc-comment.rs:3:32
   |
LL | /// doc comment with bare CR: '␍'
   |                                ^
```

Centralize the checking of unicode char width for the purposes of CLI display in one place. Account for the new replacements. Remove unneeded tracking of "zero-width" unicode chars, as we calculate these in the `SourceMap` as needed now.

estebank committed:
Replace ASCII control chars with Unicode Control Pictures

```
error: bare CR not allowed in doc-comment
  --> $DIR/lex-bare-cr-string-literal-doc-comment.rs:3:32
   |
LL | /// doc comment with bare CR: '␍'
   |                                ^
```

estebank committed:
Be more accurate about calculating `display_col` from a `BytePos`

No longer track "zero-width" chars in `SourceMap`, read directly from the line when calculating the `display_col` of a `BytePos`. Move `char_width` to `rustc_span` and use it from the emitter.

This change allows the following to properly align in terminals (depending on the font, the replaced control codepoints are rendered as 1 or 2 width, on my terminal they are rendered as 1, on VSCode text they are rendered as 2):

```
error: this file contains an unclosed delimiter
  --> $DIR/issue-68629.rs:5:17
   |
LL | ␜␟ts␀![{i
   |       -- unclosed delimiter
   |       |
   |       unclosed delimiter
LL | ␀␀  fn rݻoa>rݻm
   |                ^
```

estebank commented:
If we're concerned by legibility or by replacing with characters that might not be well supported in all terminals, we can also replace these with something like `[CR]`:

```
error: bare CR not allowed in doc-comment
  --> $DIR/lex-bare-cr-string-literal-doc-comment.rs:3:32
   |
LL | /// doc comment with bare CR: '[CR]'
   |                                ^^^^
```

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9861273312/job/27229369436) [(plain)](https://github.com/rust-lang/rust/commit/00895955d22cf576798788970b4691ba6c2fad4f/checks/27229369436/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.408   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.420 Collecting boolean-py==4.0
#13 0.423   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.436 Collecting chardet==5.1.0
---
#13 3.461 Building wheels for collected packages: reuse
#13 3.462   Building wheel for reuse (pyproject.toml): started
#13 3.787   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.788   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.788   Stored in directory: /tmp/pip-ephem-wheel-cache-w3pxmd10/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.791 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.812   Attempting uninstall: setuptools
#13 3.813     Found existing installation: setuptools 59.6.0
#13 3.814     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.814     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.814     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.478 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.478 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.989 Collecting virtualenv
#13 5.027   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.093      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 89.4 MB/s eta 0:00:00
#13 5.143 Collecting platformdirs<5,>=3.9.1
#13 5.147   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.162 Collecting distlib<1,>=0.3.7
#13 5.166   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.203 Collecting filelock<4,>=3.12.2
#13 5.206   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.206   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.294 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.466 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      214976 kB
DirectMap2M:     7124992 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/f25e92bd42b14e45440c0a30a4ed751ea502f430/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-f25e92bd42b14e45440c0a30a4ed751ea502f430-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #127777) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #127819) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
Make unicode text flow control chars visible as �

We already point these out quite aggressively, telling people not to use them, but would normally be rendered as nothing. Having them visible will make it easier for people to actually deal with them.

```
error: unicode codepoint changing visible direction of text present in literal
  --> $DIR/unicode-control-codepoints.rs:26:22
   |
LL |     println!("{:?}", '�');
   |                      ^-^
   |                      ||
   |                      |'\u{202e}'
   |                      this literal contains an invisible unicode text flow control codepoint
   |
   = note: these kind of unicode codepoints change the way text flows on applications that support them, but can cause confusion because they change the order of characters on the screen
   = help: if their presence wasn't intentional, you can remove them
help: if you want to keep them but make them visible in your source code, you can escape them
   |
LL |     println!("{:?}", '\u{202e}');
   |                       ~~~~~~~~
```

vs the previous

```
error: unicode codepoint changing visible direction of text present in literal
  --> $DIR/unicode-control-codepoints.rs:26:22
   |
LL |     println!("{:?}", '');
   |                      ^-
   |                      ||
   |                      |'\u{202e}'
   |                      this literal contains an invisible unicode text flow control codepoint
   |
   = note: these kind of unicode codepoints change the way text flows on applications that support them, but can cause confusion because they change the order of characters on the screen
   = help: if their presence wasn't intentional, you can remove them
help: if you want to keep them but make them visible in your source code, you can escape them
   |
LL |     println!("{:?}", '\u{202e}');
   |                       ~~~~~~~~
```

estebank commented on tests/ui/parser/trailing-carriage-return-in-string.stderr:
```diff
@@ -1,7 +1,7 @@
 error: unknown character escape: `\r`
   --> $DIR/trailing-carriage-return-in-string.rs:10:25
    |
-LL |     let bad = "This is \
 a test";
+LL |     let bad = "This is \␍ a test";
    |                         ^ unknown character escape
```
This is confusing, because the problem is that `\` escape is followed by a non-printable character. This is parsed as `\␍`, which is kind-of `\\r`.

oli-obk commented:
![bors r plus](https://www.pietroalbini.org/bors-r-plus.png)
<!--
@bors r+
-->

@bors rollup

bors commented:
:pushpin: Commit 9dffe9573be86ac7302fb2f27219ef8f9c78dea7 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 9dffe9573be86ac7302fb2f27219ef8f9c78dea7 -->
<!-- homu: {"type":"Approved","sha":"9dffe9573be86ac7302fb2f27219ef8f9c78dea7","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

estebank committed:
Fix ddltool-failed test

estebank commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 9bd7680b2e0f1f5680b04fdb2401bad3e082fa38 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 9bd7680b2e0f1f5680b04fdb2401bad3e082fa38 -->
<!-- homu: {"type":"Approved","sha":"9bd7680b2e0f1f5680b04fdb2401bad3e082fa38","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
@rust-timer build e3343bd96d6125c21f3e32f4be744b7c026ec6b2

rust-timer commented:
Finished benchmarking commit (e3343bd96d6125c21f3e32f4be744b7c026ec6b2): [comparison URL](https://perf.rust-lang.org/compare.html?start=e7d66eac5e8e8f60370c98d186aee9fa0ebd7845&end=e3343bd96d6125c21f3e32f4be744b7c026ec6b2&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e7d66eac5e8e8f60370c98d186aee9fa0ebd7845&end=e3343bd96d6125c21f3e32f4be744b7c026ec6b2&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8% | [0.2%, 2.9%] | 32    |
| Regressions ❌ <br /> (secondary)  | 0.5% | [0.3%, 2.2%] | 17    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.8% | [0.2%, 2.9%] | 32    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e7d66eac5e8e8f60370c98d186aee9fa0ebd7845&end=e3343bd96d6125c21f3e32f4be744b7c026ec6b2&stat=max-rss)
<details>
<summary>Results (primary -3.2%, secondary -6.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.2% | [-3.2%, -3.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -6.1% | [-6.1%, -6.1%] | 1     |
| All ❌✅ (primary)                 | -3.2% | [-3.2%, -3.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e7d66eac5e8e8f60370c98d186aee9fa0ebd7845&end=e3343bd96d6125c21f3e32f4be744b7c026ec6b2&stat=cycles:u)
<details>
<summary>Results (primary 1.2%, secondary 1.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.2% | [1.0%, 1.5%] | 3     |
| Regressions ❌ <br /> (secondary)  | 1.0% | [1.0%, 1.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.2% | [1.0%, 1.5%] | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e7d66eac5e8e8f60370c98d186aee9fa0ebd7845&end=e3343bd96d6125c21f3e32f4be744b7c026ec6b2&stat=size:linked_artifact)
<details>
<summary>Results (secondary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

**Bootstrap**: 771.74s -> 769.585s (-0.28%)
**Artifact size**: 328.91 MiB -> 328.93 MiB (0.01%)

bugadani commented:
This change seems to have a (hopefully unintended) side effect: at least rust-lld on windows seem to end lines with CRLF, so every line of the rust-lld output ends with an ugly `␍`.

estebank commented:
@bugadani I'd noticed that in one test but didn't think too much of it at the time. Yes, we should clean that up.
